### ISSUE TITLE
react-18 sample not rerendering in firefox after successfull signing using popup

### ISSUE THREAD

**Issue by fvdnabee** [AzureAD/microsoft-authentication-library-for-js#4753]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.23.0

### Wrapper Library

MSAL React (@azure/msal-react)

### Wrapper Library Version

1.3.2

### Description

The [msal-react sample for react-18](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-react-samples/react-18-sample) is not re-rendering the homepage after a successful sign-in via popup in firefox. The login is successful because after F5'ing, the homepage does login the rendered view. I can only reproduce this in firefox, in edge and chromium the homepage is rerendered after a successful sign-in. The issue only occurs when signing in via Popup, sign-in via redirect does re-render the homepage in firefox.

### Error Message

None

### Msal Logs

```
Download the React DevTools for a better development experience: https://reactjs.org/link/react-devtools [react-dom.development.js:29742](http://localhost:3000/home/fvdnabee/GIT/microsoft-authentication-library-for-js/samples/msal-react-samples/react-18-sample/node_modules/react-dom/cjs/react-dom.development.js)
[Thu, 28 Apr 2022 07:57:01 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:handleRedirectStart [authConfig.js:37](http://localhost:3000/home/fvdnabee/GIT/microsoft-authentication-library-for-js/samples/msal-react-samples/react-18-sample/src/authConfig.js)
[Thu, 28 Apr 2022 07:57:01 GMT] : [742c5e69-1e2b-49ba-91d8-e2bac8cbc1dc] : msal.js.browser@2.23.0 : Info - handleRedirectPromise called but there is no interaction in progress, returning null. authConfig.js:37
[Thu, 28 Apr 2022 07:57:01 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:handleRedirectStart results in setting inProgress from startup to handleRedirect [authConfig.js:37](http://localhost:3000/home/fvdnabee/GIT/microsoft-authentication-library-for-js/samples/msal-react-samples/react-18-sample/src/authConfig.js)
[Thu, 28 Apr 2022 07:57:01 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:handleRedirectStart results in setting inProgress from startup to handleRedirect authConfig.js:37
[Thu, 28 Apr 2022 07:57:01 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:handleRedirectEnd [authConfig.js:37](http://localhost:3000/home/fvdnabee/GIT/microsoft-authentication-library-for-js/samples/msal-react-samples/react-18-sample/src/authConfig.js)
[Thu, 28 Apr 2022 07:57:01 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:handleRedirectEnd results in setting inProgress from handleRedirect to none 2 authConfig.js:37
[Thu, 28 Apr 2022 07:57:06 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:loginStart [authConfig.js:37](http://localhost:3000/home/fvdnabee/GIT/microsoft-authentication-library-for-js/samples/msal-react-samples/react-18-sample/src/authConfig.js)
Some cookies are misusing the recommended “SameSite“ attribute 15
[Thu, 28 Apr 2022 07:57:06 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:popupOpened authConfig.js:37
[Thu, 28 Apr 2022 07:57:06 GMT] : [f821a13c-0bc6-447a-9a27-41965eaa9469] : @azure/msal-common@6.2.0 : Info - in acquireToken call [authConfig.js:37](http://localhost:3000/home/fvdnabee/GIT/microsoft-authentication-library-for-js/samples/msal-react-samples/react-18-sample/src/authConfig.js)
[Thu, 28 Apr 2022 07:57:07 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:loginSuccess authConfig.js:37
```

### MSAL Configuration

```javascript
export const msalConfig = {
    auth: {
        clientId: "753315ad-REDACTED",
        authority: "https://login.microsoftonline.com/556e6b1f-REDACTED",
        redirectUri: "/",
        postLogoutRedirectUri: "/"
    },
    cache: {
        cacheLocation: "localStorage",
        storeAuthStateInCookie: isIE || isEdge || isFirefox
    },
    system: {
        loggerOptions: {
            loggerCallback: (level, message, containsPii) => {
                if (containsPii) {	
                    return;	
                }
                switch (level) {	
                    case LogLevel.Error:	
                        console.error(message);	
                        return;	
                    case LogLevel.Info:	
                        console.info(message);	
                        return;	
                    case LogLevel.Verbose:	
                        console.debug(message);	
                        return;	
                    case LogLevel.Warning:	
                        console.warn(message);	
                        return;	
                    default:
                        return;
                }
            }
        }
    }
};
```

### Relevant Code Snippets

```javascript
Official sample is here: https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-react-samples/react-18-sample/
```

### Reproduction Steps

1. Open react-18-sample in firefox, I'm running firefox 99.0.1-1
2. Click sign in using popup
3. Provide valid credentials, the popup closes
4. The homepage does not re-render with the authenticated view

### Expected Behavior

The homepage should re-render displaying the authenticated view, manually reloading the page should not be necessary.

### Identity Provider

Azure B2C Basic Policy

### Browsers Affected (Select all that apply)

Firefox

### Source

External (Customer)

**Comment by fvdnabee** [comment]
Screencast illustrating the issue:

https://user-images.githubusercontent.com/1157936/165706727-199dcc62-ac69-44e7-9568-dfc20a8cd904.mp4

**Comment by fvdnabee** [comment]
As a comparison, the logs in chromium (where the issue does not reproduce) look as follows (for the same signin via popup in the react18 sample):
```
[Thu, 28 Apr 2022 08:01:47 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:handleRedirectStart
authConfig.js:37 [Thu, 28 Apr 2022 08:01:47 GMT] : [ff2f89b7-6c5d-4dba-9e73-01d4eec45216] : msal.js.browser@2.23.0 : Info - handleRedirectPromise called but there is no interaction in progress, returning null.
authConfig.js:37 [Thu, 28 Apr 2022 08:01:47 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:handleRedirectStart results in setting inProgress from startup to handleRedirect
authConfig.js:37 [Thu, 28 Apr 2022 08:01:47 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:handleRedirectStart results in setting inProgress from startup to handleRedirect
authConfig.js:37 [Thu, 28 Apr 2022 08:01:47 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:handleRedirectEnd
authConfig.js:37 [Thu, 28 Apr 2022 08:01:47 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:handleRedirectEnd results in setting inProgress from handleRedirect to none
authConfig.js:37 [Thu, 28 Apr 2022 08:01:47 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:handleRedirectEnd results in setting inProgress from handleRedirect to none
authConfig.js:37 [Thu, 28 Apr 2022 08:01:50 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:loginStart
authConfig.js:37 [Thu, 28 Apr 2022 08:01:50 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:loginStart results in setting inProgress from none to login
authConfig.js:37 [Thu, 28 Apr 2022 08:01:50 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:loginStart results in setting inProgress from none to login
authConfig.js:37 [Thu, 28 Apr 2022 08:01:51 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:popupOpened
authConfig.js:37 [Thu, 28 Apr 2022 08:01:51 GMT] : [dcb12b1d-7bfd-43f3-8280-61a34d3e200e] : @azure/msal-common@6.2.0 : Info - in acquireToken call
authConfig.js:37 [Thu, 28 Apr 2022 08:01:51 GMT] : @azure/msal-browser@2.23.0 : Info - Emitting event: msal:loginSuccess
authConfig.js:37 [Thu, 28 Apr 2022 08:01:51 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:loginSuccess results in setting inProgress from login to none
authConfig.js:37 [Thu, 28 Apr 2022 08:01:51 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - updating account state
authConfig.js:37 [Thu, 28 Apr 2022 08:01:51 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - msal:loginSuccess results in setting inProgress from login to none
authConfig.js:37 [Thu, 28 Apr 2022 08:01:51 GMT] : @azure/msal-react@1.3.2 : Info - MsalProvider - updating account state
```

**Comment by fvdnabee** [comment]
Are there any suggestions for a work-around for this issue? Can I force a re-render of our root react component when the pop up whenever is closed? This is blocking our project from switching to react 18.

**Comment by hectormmg** [comment]
@fvdnabee I was able to reproduce this and will be investigating to find a root cause. Once I figure out a work-around or fix I will update the issue.

**Comment by fvdnabee** [comment]
As a work-around I've tried adding react state to the Home component (which contains `AuthenticatedTemplate`) that is updated in a callback from msal-browser event's API (as per [this example](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-react/docs/events.md#registering-and-unregistering-an-event-callback)). Eventhough the event callback is executed, the state update does not appear to trigger a re-render of the Home component. Not sure what is blocking the render.

Patch to Home.jsx is as follows:
```diff
diff --git a/samples/msal-react-samples/react-18-sample/src/pages/Home.jsx b/samples/msal-react-samples/react-18-sample/src/pages/Home.jsx
index 288c3eb8b..e1c049a8b 100644
--- a/samples/msal-react-samples/react-18-sample/src/pages/Home.jsx
+++ b/samples/msal-react-samples/react-18-sample/src/pages/Home.jsx
@@ -3,10 +3,38 @@ import Button from "@mui/material/Button";
 import ButtonGroup from "@mui/material/ButtonGroup";
 import Typography from "@mui/material/Typography";
 import { Link as RouterLink } from "react-router-dom";
+import { useEffect, useState } from "react";
+import { useMsal } from "@azure/msal-react";
+import { EventType } from "@azure/msal-browser";
 
 export function Home() {
+    const { instance } = useMsal();
+    const [value, setValue] = useState(0)
+
+    useEffect(() => {
+        // This will be run on component mount
+        const callbackId = instance.addEventCallback((message) => {
+            // This will be run every time an event is emitted after registering this callback
+            if (message.eventType === EventType.LOGIN_SUCCESS) {
+                // const result = message.payload;
+                // Do something with the result
+                setValue(v => v + 1);
+                console.log("Hello from callback");
+            }
+        });
+
+        return () => {
+            // This will be run on component unmount
+            if (callbackId) {
+                instance.removeEventCallback(callbackId);
+            }
+        }
+
+    }, [instance]);
+    console.log(value);
   return (
       <>
+          {value}
```

**Comment by fvdnabee** [comment]
Issue is still present after updating to msal-react 1.4.0 & msal-browser 2.24.0.

**Comment by tnorling** [comment]
@fvdnabee For some reason the moment the popup is opened the whole UI freezes and nothing is clickable anymore. It's not clear to me what is causing this yet but I have found a way to fix it. Setting the `asyncPopups` flag to `true` on [the `PublicClientApplication` config](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/configuration.md#system-config-options) seems to resolve the issue for me, can you try that and let me know if that works for you as well? In the meantime I'll continue to investigate why this is happening with the default setting.

**Comment by fvdnabee** [comment]
`asyncPopups: true` is indeed a work-around for this issue, thanks! Do you know if `asyncPopups` has any down sides (i.e. why is the default false)?

**Comment by tnorling** [comment]
Async popups are potentially more likely to be blocked by popup blockers

**Comment by himadrinath** [comment]
same problem seeing in firefox and safari

**Comment by himadrinath** [comment]
looks like problem has been fixed in react@latest release

**Comment by tnorling** [comment]
Thanks for the update @himadrinath! @fvdnabee Does upgrading react resolve this issue for you as well?

**Comment by omegaji** [comment]
for me updating react from 18.1.0 to 18.2.0 did the job! Thank you all!

**Comment by tnorling** [comment]
Closing as it sounds like this has been resolved by upgrading React. Please open a new issue if you're still having issues.
